text/3502-cargo-script.md

```diff
@@ -0,0 +1,1205 @@
+  - This may be altered with [rust-lang/cargo#9829](https://github.com/rust-lang/cargo/issues/9829)
+- `package.publish = false` to avoid accidental publishes, particularly if we
+  later add support for including them in a workspace.
+- `package.edition = <current>` to avoid always having to add an embedded
+  manifest at the cost of potentially breaking scripts on rust upgrades
+  - Warn when `edition` is unspecified.  While with single-file packages this will be
+    silenced by default, users wanting stability are also likely to be using
+    other commands, like `cargo test` and will see it.
+  - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand
+  - Based on feedback, we can switch to "edition is required as of <future> edition"
+
+Disallowed manifest fields:
+- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
+- `package.workspace`, `package.build`, `package.links`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`
+
+Single-file packages maintain an out-of-line target directory by default.
+This is `$CARGO_HOME/target/<hash-of-path>`.
+
+A single-file package is accepted by cargo commands as a `--manifest-path`
+- This is distinguished by the file extension (`.rs`) and that it is a file.
+- This allows running `cargo test --manifest-path single.rs`
+- `cargo package` / `cargo publish` will normalize this into a multi-file package
+- `cargo add` and `cargo remove` may not support editing embedded manifests initially
+- Path-dependencies may not refer to single-file packages at this time (they don't have a `lib` target anyways)
+
+Single-file packages will not be accepted as `path` or `git` dependencies.
+
+The lockfile for single-file packages will be placed in `CARGO_TARGET_DIR`.  In
+the future, when workspaces are supported, that will allow a user to have a
+persistent lockfile.
+We may also allow customizing the non-workspace lockfile location in the [future](#future-possibilities).
+
+## `cargo <file>.rs`
+
+`cargo` is intended for putting in the `#!` for single-file packages:
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+    println!("Hello world");
+}
+```
+- Like with `cargo install`, `.cargo/config.toml` will be read based on the
+  scripts location rather than the current-dir.
+  - And like `cargo install`, the current-dir rustup-toolchain is respected
+- `--release` is not passed in because the primary use case is for exploratory
+  programming, so the emphasis will be on build-time performance and debugging,
+  rather than runtime performance
+
+Most other flags and behavior will be similar to `cargo run`.
+
+The precedence for `cargo foo` will change from:
+1. built-in commands
+2. user aliases
+3. third-party commands
+
+to:
+1. built-in command xor manifest
+2. user aliases
+3. third-party commands
+
+To allow the xor, we enforce that
+- manifests must be passed in as `Cargo.toml`, `foo.rs`, or have a `/` in them
```

Comment:
And may not have a `.` in the filename unless they end in `.rs`:

```console
> cat test.hello
#!/usr/bin/env -S cargo -Zscript

fn main() {
    println!("hello world");
}
> ./test.hello
error: the manifest-path must be a path to a Cargo.toml file
```

(assuming the current implementation has the intended semantics here)

Reply:
It says it must be `foo.rs`.  I've clarified that by shifting from an example to saying "with a `.rs` extension".

Reply:
It still reads to me as if ‘./hello.test‘ would be allowed because it matches the third clause, but it's not because only extensionless filenames are checked for ‘/‘.

Reply:
I don't personally use an iPad so I can't verify that it's impossible but last I checked if you hold `'` for a second or so it'll pop up with alternative quote options.

It'll look like so:
![Quote options on Apple Keyboard](https://github.com/rust-lang/rfcs/assets/109059814/d3a37ce4-da1f-4e03-aa10-e0c820935fcb)

Reply:
I see. This is specifically referring to disambiguating `cargo foo`.

When not on mobile, ill need to check if there is another place where we talk about how to parse the file.

Reply:
a4763f18 updated the section on this matching the current implementation in cargo.